002-es6-modules.md

```diff
@@ -0,0 +1,274 @@
+	1. Evaluate immediately
+	2. Produce a DynamicModuleRecord from `module.exports`
+3. If ES6
+	1. Parse for `import`/`export`s and keep record, in order to create bindings
+	2. Gather all submodules by performing `require` recursively
+		* See circular dep semantics below 
+	3. Connect `import` bindings for all relevant submodules (see [ModuleDeclarationInstantiation](https://tc39.github.io/ecma262/#sec-moduledeclarationinstantiation))
+	4. Evaluate
+
+## Semantics
+
+### Determining if source is an ES6 Module
+
+If a module can be parsed outside of the ES6 Module goal, it will be treated as a [Script](https://tc39.github.io/ecma262/#sec-parse-script). Otherwise it will be parsed as [ES6](https://tc39.github.io/ecma262/#sec-parsemodule).
+
+In pseudocode, it looks somewhat like:
+
+```
+try {
+  v8::Script::Compile
+}
+catch (e) {
+  v8::Script::CompileModule
+}
+```
+
+V8 may or may not choose to use parser fallback to combine this into one step.
+
+### CommonJS consuming ES6
+
+#### default exports
+
+ES6 modules only ever declare named exports. A default export just exports a property named `default`.
+
+Given
+
+```javascript
+let foo = 'my-default';
+export default foo;
+```
+
+```javascript
+require('./es6');
+// {default:'my-default'}
+```
+
+#### read only
+
+The objects create by an ES6 module are [ModuleNamespace Objects](https://tc39.github.io/ecma262/#sec-module-namespace-exotic-objects).
+
+These have `[[Set]]` be a no-op and are read only views of the exports of an ES6 module.
+
+### ES6 consuming CommonJS
+
+#### default imports
+
+`module.exports` shadows `module.exports.default`.
+
+This means that if there is a `.default` on your CommonJS exports it will be shadowed to the `module.exports` of your module.
+
+Given:
+
+```javascript
+// cjs.js
+module.exports = {default:'my-default'};
+```
+
+You will grab `module.exports` when performing an ES6 import.
+
+```javascript
+// es6.js
+import foo from './cjs';
+// foo = {default:'my-default'};
+
+import {default as bar} from './cjs';
+// bar = {default:'my-default'};
+```
+
+### Known Gotchas
+
+All of these gotchas relate to opt-in semantics and the fact that CommonJS is a dynamic loader while ES6 is a static loader.
+
+No existing code will be affected.
+
+#### Circular Dep CJS => ES6 => CJS
+
+Given:
+
+```javascript
+//cjs.js
+module.exports = {x:0};
+require('./es6');
+```
+
+```javascript
+//es6.js
+import * as ns from './cjs';
+// ns = undefined
+import cjs from './cjs';
+// ns = undefined
+```
+
+The list of properties being exported is not populated until `cjs.js` finishes executing.
+
+The result is that there are no properties to import and you recieve an empty module.
+
+##### Option: Throw on this case
+
+Since this case is coming from ES6, we could throw whenever this occurs. This is similar to how you cannot change a generator while it is running.
+
+If taken this would change the ES6 module behavior to:
+
+```javascript
+//es6.js
+import * as ns from './cjs';
```

Comment:
While the requirement to have file extensions may be feasible within the same module, after some thought I don't think it would work once we hit `node_modules`.

In particular these situations would no longer work as dependencies:

* submodule uses `index.js` and does not have a main in package.json
* submodule package.json main does not list a file extension
* `node_modules/foo.js`

Also, the current resolution mechanic for the module loader is compatible with the Browser proposal (all forms of the browser resolution are represented by it). Unless it is incompatible in some way I don't see.

Is there a particular feature to be gained by this limit? ES6 modules give us many things by sticking to static lists of import/export names, I am guessing there is something here.